### ISSUE TITLE
Issues during setup docker and jupyter

### ISSUE THREAD

**Issue by wanderjones** [deepfakes/faceswap-playground#245]
I am admittedly a complete noob, but if anyone could help it would be very appreciated!

## Expected behavior

Trying to mount the faceswap volume for a CPU only using docker on mac.

## Actual behavior

Will get this error everytime
docker: Error response from daemon: OCI runtime create failed: container_linux.go:344: starting container process caused "exec: "/run_jupyter.sh": stat /run_jupyter.sh: no such file or directory": unknown. ERRO[0001] error waiting for container: context canceled

## Steps to reproduce

Running as per the instructions in the setup.py

## Other relevant information
- **Command lined used (if not specified in steps to reproduce)**: faceswap.py ...

docker run -p 8888:8888     --hostname deepfakes-cpu --name deepfakes-cpu     -v /Users/USER/Desktop/OldMBP/faceswap:/srv     deepfakes-cpu

- **Operating system and version:** Windows, macOS, Linux 
MacOS 10.14.2
- **Python version:** 2.7, 3.5, 3.6.4, ...
Python 3.5.4rc1
- **Faceswap version:** commit hash or version number
Not sure.
- **Faceswap method:** CPU/GPU
CPU

**Comment by torzdf** [comment]
Unfortunately I don't use Docker. Maybe this may help:
https://github.com/deepfakes/faceswap-playground/issues/234

Otherwise @ DKingCN  might have some ideas?

**Comment by DKingCN** [comment]
well, maybe the entrypoint run_jupyter.sh is not there anymore.
as it was set in Dockerfile.cpu
```
WORKDIR "/notebooks"
CMD ["/run_jupyter.sh", "--allow-root"]
```

Try `docker run -p 8888:8888 --hostname deepfakes-cpu --name deepfakes-cpu -v /Users/USER/Desktop/OldMBP/faceswap:/srv deepfakes-cpu /bin/bash`. This will get rid of jupyter thing and take you straight to a bash shell.

If you are interested, please find us a new entrypoint you believe its good enough. it will be helpful.

**Comment by GideonWeng** [comment]
> well, maybe the entrypoint run_jupyter.sh is not there anymore.
> as it was set in Dockerfile.cpu
> 
> ```
> WORKDIR "/notebooks"
> CMD ["/run_jupyter.sh", "--allow-root"]
> ```
> Try `docker run -p 8888:8888 --hostname deepfakes-cpu --name deepfakes-cpu -v /Users/USER/Desktop/OldMBP/faceswap:/srv deepfakes-cpu /bin/bash`. This will get rid of jupyter thing and take you straight to a bash shell.
> 
> If you are interested, please find us a new entrypoint you believe its good enough. it will be helpful.

I met the same problem as @wanderjones on my server. Following is **Other relevant information** of mine:  
* **Command lined used (if not specified in steps to reproduce)**: faceswap.py ...
```sh
docker run -p 8888:8888 --hostname deepfakes-cpu --name deepfakes-cpu -v /home/turtleman/work_space/face-play/faceswap:/srv deepfakes-cpu
```

* **Operating system and version:** Windows, macOS, Linux
  Linux-4.9.0-8-amd64-x86_64-with-debian-9.7
* **Python version:** 2.7, 3.5, 3.6.4, ...
  Python 3.6.8
* **Faceswap version:** commit hash or version number
  e37c2ef4743cd9ab79d0cd2a503586b767f5692a
* **Faceswap method:** 
  CPU   
Then I choose your suggestion, run the command:   
```sh 
docker run -p 8888:8888 --hostname deepfakes-cpu --name deepfakes-cpu -v /home/turtleman/work_space/face-play/faceswap:/srv deepfakes-cpu /bin/bash
```
A new strange thing comes, the docker container exits immediately after I start it. I've tried to find solutions in the internet, still not solved. Appreciate any help!

**Comment by DKingCN** [comment]
i will look into this soon. please be patient.

**Comment by DKingCN** [comment]
OK, basically the problem is at the `docker run` command line.

As before jupyter existed as a foreground entrypoint, `docker run` will work without `-ti`.
Since jupyter is removed from tensorflow:latest-py3, `docker run ... bash` will exit at once without `-ti` or `-tid`. 

I will NOT explain the meaning of `-tid` because you should have read it in docker documentations.

So the fix is to add `-tid` when run the image.

The fix has been updated(force-pushed) in my PR in deepfakes.

**Comment by GideonWeng** [comment]
> OK, basically the problem is at the `docker run` command line.
> 
> As before jupyter existed as a foreground entrypoint, `docker run` will work without `-ti`.
> Since jupyter is removed from tensorflow:latest-py3, `docker run ... bash` will exit at once without `-ti` or `-tid`.
> 
> I will NOT explain the meaning of `-tid` because you should have read it in docker documentations.
> 
> So the fix is to add `-tid` when run the image.
> 
> The fix has been updated(force-pushed) in my PR in deepfakes.

Thanks a lot! I've entered the container.

**Comment by GideonWeng** [comment]
> OK, basically the problem is at the `docker run` command line.
> 
> As before jupyter existed as a foreground entrypoint, `docker run` will work without `-ti`.
> Since jupyter is removed from tensorflow:latest-py3, `docker run ... bash` will exit at once without `-ti` or `-tid`.
> 
> I will NOT explain the meaning of `-tid` because you should have read it in docker documentations.
> 
> So the fix is to add `-tid` when run the image.
> 
> The fix has been updated(force-pushed) in my PR in deepfakes.

But why that container couldn't work well?  
When building the image, a warning occurred:
```sh
/usr/local/lib/python3.5/dist-packages/pip/_internal/commands/install.py:207: UserWarning: Disabling all use of wheels due to the use of --build-options / --global-options / --install-options.
  cmdoptions.check_install_build_global(options)
```
I ignored it.   
After entering the container and running the command `root@deepfakes-cpu:/srv# python faceswap.py extract -i ./faceswap-data_gathering/photos/trump/ -o ./data/trump`. Error:  
```sh
Traceback (most recent call last):
  File "faceswap.py", line 5, in <module>
    import lib.cli as cli
  File "/srv/lib/cli.py", line 12, in <module>
    from lib.utils import safe_shutdown
  File "/srv/lib/utils.py", line 13, in <module>
    import cv2
  File "/usr/local/lib/python3.5/dist-packages/cv2/__init__.py", line 3, in <module>
    from .cv2 import *
ImportError: libgthread-2.0.so.0: cannot open shared object file: No such file or directory
```
Thus I run `apt-get install libglib2.0-0`.   
Then run `python faceswap.py extract -i ./faceswap-data_gathering/photos/trump/ -o ./data/trump`. Then same error(as without docker I ran before)comes:
```sh
Traceback (most recent call last):
  File "/srv/lib/cli.py", line 90, in execute_script
    process.process()
  File "/srv/scripts/extract.py", line 49, in process
    self.run_extraction()
  File "/srv/scripts/extract.py", line 152, in run_extraction
    desc="Extracting faces"):
  File "/usr/local/lib/python3.5/dist-packages/tqdm/_tqdm.py", line 1022, in __iter__
    for obj in iterable:
  File "/srv/scripts/extract.py", line 412, in detect_faces
    raise Exception(err)
Exception: Error in child process 1259. Traceback (most recent call last):
  File "/srv/plugins/extract/align/_base.py", line 91, in run
    self.align(*args, **kwargs)
  File "/srv/plugins/extract/align/fan.py", line 56, in align
    super().align(*args, **kwargs)
  File "/srv/plugins/extract/align/_base.py", line 81, in align
    self.initialize(*args, **kwargs)
  File "/srv/plugins/extract/align/fan.py", line 40, in initialize
    raise ValueError("No Graphics Card Detected! FAN is not currently supported on CPU. "
ValueError: No Graphics Card Detected! FAN is not currently supported on CPU. Use another aligner.
``` 
The crash report([crash_report.2019.02.10.104421485944.log](https://github.com/deepfakes/faceswap-playground/files/2848827/crash_report.2019.02.10.104421485944.log)) in case you need.

**Comment by wanderjones** [comment]
I get 

> Traceback (most recent call last):
  File "faceswap.py", line 5, in <module>
    import lib.cli as cli
  File "/srv/lib/cli.py", line 12, in <module>
    from lib.utils import safe_shutdown
  File "/srv/lib/utils.py", line 13, in <module>
    import cv2
  File "/usr/local/lib/python3.5/dist-packages/cv2/__init__.py", line 3, in <module>
    from .cv2 import *
ImportError: libgthread-2.0.so.0: cannot open shared object file: No such file or directory

As well, but on a mac, and I am not sure on how to progress.

**Comment by GideonWeng** [comment]
> I get
> 
> > Traceback (most recent call last):
> > File "faceswap.py", line 5, in 
> > import lib.cli as cli
> > File "/srv/lib/cli.py", line 12, in 
> > from lib.utils import safe_shutdown
> > File "/srv/lib/utils.py", line 13, in 
> > import cv2
> > File "/usr/local/lib/python3.5/dist-packages/cv2/**init**.py", line 3, in 
> > from .cv2 import *
> > ImportError: libgthread-2.0.so.0: cannot open shared object file: No such file or directory
> 
> As well, but on a mac, and I am not sure on how to progress.

I solved this by running `apt-get install libglib2.0-0`(maybe `sudo` needed in ur pc). But I still happened to another issue afterwards.

**Comment by wanderjones** [comment]
apt-get is a linux command, doesn't work on mac.

**Comment by torzdf** [comment]
https://www.google.com/search?safe=off&client=firefox-b-d&ei=dmdhXKegC8OV1fAPoJ6smA4&q=libglib2+mac&oq=libglib2+mac&gs_l=psy-ab.3..33i160.1788.1908..2094...0.0..0.93.154.2......0....1..gws-wiz.2BRSnXUizYg

**Comment by stefanondisponibile** [comment]
I solved it on **Ubuntu 18.04** by applying the following updates to the _Dockerfile.gpu_:

`@1`
**-**  `FROM tensorflow/tensorflow:latest-gpu-py3`
**+** `FROM tensorflow/tensorflow:latest-gpu-py3-jupyter`

Comment out the last two lines:
**-** `WORKDIR "/"`
**-** `CMD ["jupyter", "notebook", "$@", "--allow-root"]`

**+** `# WORKDIR "/"`
**+** `# CMD ["jupyter", "notebook", "$@", "--allow-root"]`
